lib/Dialect/FIRRTL/Transforms/ExpandWhens.cpp

```diff
@@ -22,89 +21,71 @@
+protected:
+  /// Tracks if anything in the IR has changed.
+  bool changed = false;
 
-  /// Run expand whens on the Module.  This will emit an error for each
-  /// incomplete initialization found. If an initialiazation error was detected,
-  /// this will return failure and leave the IR in an inconsistent state.
-  static LogicalResult run(FModuleOp op);
-
-  using FIRRTLVisitor<ExpandWhensVisitor>::visitExpr;
-  using FIRRTLVisitor<ExpandWhensVisitor>::visitDecl;
-  using FIRRTLVisitor<ExpandWhensVisitor>::visitStmt;
-
-  void visitDecl(FModuleOp op);
-  void visitDecl(InstanceOp op);
-  void visitDecl(MemOp op);
-  void visitDecl(RegOp op);
-  void visitDecl(RegResetOp op);
-  void visitDecl(WireOp op);
-  void visitInvalidOp(Operation *op) {}
-  void visitStmt(AssertOp op);
-  void visitStmt(AssumeOp op);
-  void visitStmt(ConnectOp op);
-  void visitStmt(CoverOp op);
-  void visitStmt(ModuleOp op);
-  void visitStmt(PartialConnectOp op);
-  void visitStmt(PrintFOp op);
-  void visitStmt(StopOp op);
-  void visitStmt(WhenOp op);
-  void visitUnhandledOp(Operation *op) {}
+  /// Map of destinations and the operation which is driving a value to it in
+  /// the current scope. This is used for resolving last connect semantics, and
+  /// for retrieving the responsible connect operation.
+  ScopeMap &scope;
 
-private:
-  /// Process a block, recording each declaration, and expanding all whens.
-  void process(Block &block);
+public:
+  LastConnectResolver(ScopeMap &scope) : scope(scope) {}
+
+  using FIRRTLVisitor<ConcreteT>::visitExpr;
+  using FIRRTLVisitor<ConcreteT>::visitDecl;
+  using FIRRTLVisitor<ConcreteT>::visitStmt;
 
   /// Records a connection to a destination. This will delete a previous
   /// connection to a destination if there was one.
-  void setLastConnect(FieldRef dest, Operation *connection) {
+  bool setLastConnect(FieldRef dest, Operation *connection) {
     // Try to insert, if it doesn't insert, replace the previous value.
     auto itAndInserted = scope.insert({dest, connection});
     if (!std::get<1>(itAndInserted)) {
       auto iterator = std::get<0>(itAndInserted);
       // Delete the old connection if it exists. Null connections are inserted
       // on declarations.
-      if (auto *oldConnect = iterator->second)
+      if (auto *oldConnect = iterator->second) {
         oldConnect->erase();
+      }
       iterator->second = connection;
+      return true;
     }
+    return false;
   }
 
   /// Get the destination value from a connection.  This supports any operation
   /// which is capable of driving a value.
   static Value getDestinationValue(Operation *op) {
-    return TypeSwitch<Operation *, Value>(op)
-        .Case<ConnectOp>([](auto op) { return op.dest(); })
-        .Case<PartialConnectOp>([](auto op) { return op.dest(); })
-        .Default([](Operation *op) {
-          llvm_unreachable("unknown operation");
-          return nullptr;
-        });
+    return cast<ConnectOp>(op).dest();
```

Comment:
Nice. I'm glad we can remove this since lower types guarantees the removal of Partial Connect.